maze_dataset/tokenization/maze_tokenizer.py

```diff
@@ -901,122 +1768,163 @@ def _sequence_tokens(
-            self.prompt_sequencer,
-        ]
+    def tokenizer_elements(self):
+        return [self.prompt_sequencer, *self.prompt_sequencer.tokenizer_elements()]
+
+    def tokenizer_element_tree(self, abstract: bool = False):
+        """
+        Returns a string representation of the tree of tokenizer elements contained in `self`.
+
+        # Parameters
+        - `abstract: bool`: Whether to print the name of the abstract base class or the concrete class for each `TokenizerElement` instance.
+        """
+
+        return "\n".join(
+            [
+                type(self).__name__,
+                self.prompt_sequencer.tokenizer_element_tree(
+                    abstract=abstract, depth=1
+                ),
+            ]
+        )
 
     @property
     def name(self) -> str:
         """Serializes MazeTokenizer into a key for encoding in zanj"""
-        return repr(self)
+        return "-".join([type(self).__name__, self.prompt_sequencer.name])
 
-    def summary(self) -> dict[str, Any]:
+    def summary(self) -> dict[str, TokenizerElement]:
         return {
-            name: getattr(self, name)
-            for name in [
-                "prompt_sequencer",
-                "coord_tokenizer",
-                "adj_list_tokenizer",
-                "target_tokenizer",
-                "path_tokenizer",
-            ]
+            # "prompt_sequencer": self.prompt_sequencer.name,
+            **{elem.attribute_key(): elem for elem in self.tokenizer_elements}
         }
 
-    @classmethod
-    def from_name(cls, key: str) -> "MazeTokenizer2":
-        """Builds a MazeTokenizer from the output of `MazeTokenizer2.name`"""
-        raise NotImplementedError
+    def has_element(
+        self,
+        elements: (
+            type[TokenizerElement]
+            | TokenizerElement
+            | Iterable[type[TokenizerElement] | TokenizerElement]
+        ),
+    ) -> bool:
+        """Returns True if the `MazeTokenizerModular` instance contains ALL of the items specified in `elements`.
+
+        Querying with a partial subset of `TokenizerElement` fields is not currently supported.
+        To do such a query, assemble multiple calls to `has_elements`.
+
+        # Parameters
+        - `elements`: Singleton or iterable of `TokenizerElement` instances or classes.
+        If an instance is provided, then comparison is done via instance equality.
+        If a class is provided, then comparison isdone via `isinstance`. I.e., any instance of that class is accepted.
+        """
+
+        def type_check(obj: any) -> None:
+            if not (
+                isinstance(obj, TokenizerElement)
+                or (isinstance(obj, type) and issubclass(obj, TokenizerElement))
+            ):
+                raise TypeError(
+                    f"{elements} is not a `TokenizerElement` instance or subclass."
+                )
+
+        def has_element_singular(el: type[TokenizerElement] | TokenizerElement):
```

Comment:
this and the function above should not be defined inside this function (again for parallelism reasons)